src/metabase/models/pulse_channel.clj

```diff
@@ -0,0 +1,173 @@
+
+   {:email {:name \"Email\", :recipients? true}
+    :slack {:name \"Slack\", :recipients? false}}"
+  {:email {:recipients? true}
+   :slack {:recipients? false}})
+
+(defn channel-type?
+  "Predicate function which returns `true` if the given argument is a valid value as a channel-type, `false` otherwise."
+  [channel-type]
+  (contains? (set (keys channel-types)) (keyword channel-type)))
+
+(def schedule-types
+  "Map which contains the definitions for each type of pulse schedule type we allow.  Each key is a schedule-type with
+   a map which contains any other relevant information related to the defined schedule-type.  E.g.
+
+   {:hourly {:name \"Hourly\"}
+    :dailye {:name \"Daily\"}}"
+  {:hourly {}
+   :daily  {}
+   :weekly {}})
+
+(defn schedule-type?
+  "Predicate function which returns `true` if the given argument is a valid value as a schedule-type, `false` otherwise."
+  [schedule-type]
+  (contains? (set (keys schedule-types)) (keyword schedule-type)))
+
+(defn supports-recipients?
+  "Predicate function which returns `true` if the given channel type supports a list of recipients, `false` otherwise."
+  [channel]
+  (boolean (:recipients? (get channel-types (keyword channel)))))
+
+
+(def days-of-week
+  "Simple `vector` of the days in the week used for reference and lookups.
+
+   NOTE: order is important here!!
+         these indexes match the values from clj-time `day-of-week` function (0 = Sunday, 6 = Saturday)"
+  [{:id "sun" :name "Sun"},
+   {:id "mon" :name "Mon"},
+   {:id "tue" :name "Tue"},
+   {:id "wed" :name "Wed"},
+   {:id "thu" :name "Thu"},
+   {:id "fri" :name "Fri"},
+   {:id "sat" :name "Sat"}])
+
+(def times-of-day
+  [{:id "morning" :name "Morning" :realhour 8},
+   {:id "midday" :name "Midday" :realhour 12},
+   {:id "afternoon" :name "Afternoon" :realhour 16},
+   {:id "evening" :name "Evening" :realhour 20},
+   {:id "midnight" :name "Midnight" :realhour 0}])
+
+(defn time-of-day->realhour
+  "Time-of-day to realhour"
+  [time-of-day]
+  (-> (filter #(= time-of-day (:id %)) times-of-day)
+      first
+      :realhour))
+
+
+;; ## Entity
+
+(defrecord PulseChannelInstance []
+  ;; preserve normal IFn behavior so things like ((sel :one Database) :id) work correctly
+  clojure.lang.IFn
+  (invoke [this k]
+    (get this k)))
+
+(extend-ICanReadWrite PulseChannelInstance :read :always, :write :superuser)
+
+(defentity PulseChannel
+  [(k/table :pulse_channel)
+   (hydration-keys pulse_channel)
+   (types :details :json, :schedule_details :json)
+   timestamped]
+
+  (post-select [_ {:keys [id creator_id details] :as pulse-channel}]
+    (map->PulseChannelInstance
+      (u/assoc* pulse-channel
+                ;; don't include `:emails`, we use that purely internally
+                :details    (dissoc details :emails)
+                ;; here we recombine user details w/ freeform emails
+                :recipients (delay (into (or (->> (:emails details)
```

Comment:
Instead of 
```clojure
(or (->> (:emails details)
         (mapv #(assoc {} :email %)))
    [])
```

I would go with
```clojure
(mapv (partial (array-map :email))
      (:emails details))
```

*  `mapv` will return an empty vector in cases where `(:emails details)` is `nil`, so you don't need the `or`
*  I think using the `->>` threading macro here defeats its purpose since you're only threading through a single form. It doesn't add anything to the code's readability
*  It's preferable to use `partial` over `#()` functions whenever possible. `#()` functions can't be nested amongst other shortcomings
*  Probably better to create your maps all at once with `array-map` instead of first creating an empty map with `{}` and then creating another new one that includes `:email`